### PR TITLE
SI-8151 Remove -Yself-in-annots and associated implementation

### PR DESCRIPTION
**Pull request by retronym**
This experimental option typechecked arguments of annotations
with an injected value in scope named `self`:

```
@Foo(self.foo < 1)
```

This has been slated for removal [1] for some time.

This commit removes it in one fell swoop, without any attempt
at source compatibility with code that constructs or pattern
matches on AnnotatedType.

[1] https://groups.google.com/d/msg/scala-internals/VdZ5UJwQFGI/C6tZ493Yxx4J

### CODE REVIEW

**Comment by retronym** [comment]
Review by @gkossakowski @xeno-by

File: src/reflect/scala/reflect/internal/pickling/UnPickler.scala [hunk 1]
```diff
@@ -362,7 +362,7 @@ abstract class UnPickler {
         case METHODtpe                 => MethodTypeRef(readTypeRef(), readSymbols())
         case POLYtpe                   => PolyOrNullaryType(readTypeRef(), readSymbols())
         case EXISTENTIALtpe            => ExistentialType(underlying = readTypeRef(), quantified = readSymbols())
-        case ANNOTATEDtpe              => AnnotatedType(underlying = readTypeRef(), annotations = readAnnots(), selfsym = NoSymbol)
```
Comment (gkossakowski) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala:365]: Do I read it right that pickler would pickle the symbol of `self` (when `-Yself-in-annots` is enabled) but unpickler would ignore it completely anyway?

**Comment by gkossakowski** [comment]
The failure is spurious caused by broken IDE pr validator.

File: src/reflect/scala/reflect/internal/pickling/UnPickler.scala (hunk 1 above)
Comment (retronym) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala:365]: That's how I read it, too.

Comment (gkossakowski) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala:365]: Ok, so this was broken for separate compilation which means it's even safer to just remove this without any deprecation process.

**Comment by retronym** [comment]
No, its a legitimate error.

```
[ERROR] /localhome/jenkins/b/workspace/pr-scala-integrate-ide/target/scala-refactoring/org.scala-refactoring.library/src/main/scala/scala/tools/refactoring/common/TreeTraverser.scala:157: error: wrong number of arguments for pattern TreeTraverser.this.global.AnnotatedType(annotations: List[TreeTraverser.this.global.AnnotationInfo],underlying: TreeTraverser.this.global.Type)
[ERROR]             case (AnnotatedType(annotations, underlying, selfsym), _) =>
[ERROR]                                ^
[ERROR] one error found
```

**Comment by xeno-by** [comment]
A surprising amount of simplification! I especially liked -50 lines in typer. LGTM.

**Comment by soc** [comment]
This LGTM.

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by gkossakowski** [comment]
@retronym: I'm wondering if we could defer api changes to `AnnotatedType` but merge the rest? We could just always return `NoSymbol` for `selfsym` for now. WDYT?

**Comment by retronym** [comment]
What about the `unapply`? I don't see a clean way to leave that signature alone in the reflection API but remove the field from the case class in the `internal` layer of the cake.

**Comment by retronym** [comment]
Oh, I guess you're proposing a way to fix the build error in Scala IDE?

I've just submitted a PR to fix that. It will remain compatible with 2.10.x and 2.11.0-M7, so we don't need a coordinated merge.

https://github.com/scala-ide/scala-ide/pull/627

**Comment by skyluc** [comment]
The change in Scala IDE has been merged.

**Comment by retronym** [comment]
PLS REBUILD ALL

Thanks, Luc.

**Comment by xeno-by** [comment]
The commit is green. Shall we merge?

**Comment by retronym** [comment]
I'll leave that to @gkossakowski

**Comment by gkossakowski** [comment]
LGTM.

I had an impression that IDE team is not fond of coordinated changes like that. I was wrong.

**Comment by retronym** [comment]
It wasn't exactly a coordinated change. I just changed the IDE (and scala-refactoring) in a way that is source compatible with the compiler before _and_ after this patch.

**Pull request merged**